File: lib-database/src/main/java/com/ustadmobile/core/db/dao/DashboardEntryDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmRepository;
import com.ustadmobile.lib.database.annotation.UmUpdate;
import com.ustadmobile.lib.db.entities.DashboardEntry;
import com.ustadmobile.lib.db.sync.dao.SyncableDao;

@UmDao(updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN, insertPermissionCondition =
        RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
public abstract class DashboardEntryDao
        implements SyncableDao<DashboardEntry, DashboardEntryDao> {


    @UmQuery("UPDATE DashboardEntry SET dashboardEntryTitle = :title " +
            " WHERE dashboardEntryUid = :uid")
    public abstract void updateTitle(long uid, String title, UmCallback<Integer> resultCallback);

    @UmQuery("SELECT * FROM DashboardEntry WHERE " +
            "dashboardEntryPersonUid = :uid AND dashboardEntryActive = 1 ORDER BY dashboardEntryIndex ASC")
    public abstract UmProvider<DashboardEntry> findByPersonAndActiveProvider(long uid);


    @UmQuery("UPDATE DashboardEntry SET dashboardEntryIndex = -1 WHERE  dashboardEntryUid = :uid")
    public abstract void pinEntry(long uid, UmCallback<Integer> resultCallback);

    @UmQuery("UPDATE DashboardEntry SET dashboardEntryIndex = " +
            "(SELECT (SELECT MAX(dashboardEntryIndex) FROM DashboardEntry) +1) " +
            "WHERE dashboardEntryUid = :uid")
    public abstract void unpinEntry(long uid, UmCallback<Integer> resultCallback);

    @UmQuery("UPDATE DashboardEntry SET dashboardEntryActive = 0 WHERE dashboardEntryUid = :uid")
    public abstract void deleteEntry(long uid, UmCallback<Integer> resultCallback);

    @UmQuery("SELECT * FROM DashboardEntry WHERE dashboardEntryUid = :uid")
    public abstract void findByUidAsync(long uid, UmCallback<DashboardEntry> resultCallback);

    @UmUpdate
    public abstract void updateAsync(DashboardEntry entity, UmCallback<Integer> resultCallback);

}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/DashboardEntryDao.kt
package com.ustadmobile.core.db.dao

import androidx.paging.DataSource
import androidx.room.Dao
import androidx.room.Query
import androidx.room.Update
import com.ustadmobile.lib.database.annotation.UmDao
import com.ustadmobile.lib.database.annotation.UmRepository
import com.ustadmobile.lib.db.entities.DashboardEntry

@UmDao(updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN,
        insertPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@Dao
@UmRepository
abstract class DashboardEntryDao : BaseDao<DashboardEntry> {


    @Query("UPDATE DashboardEntry SET dashboardEntryTitle = :title " +
            " WHERE dashboardEntryUid = :uid")
    abstract suspend fun updateTitle(uid: Long, title: String):Int

    @Query("SELECT * FROM DashboardEntry WHERE " + "dashboardEntryPersonUid = :uid " +
            " AND CAST(dashboardEntryActive AS INTEGER) = 1 ORDER BY dashboardEntryIndex ASC")
    abstract fun findByPersonAndActiveProvider(uid: Long): DataSource.Factory<Int,DashboardEntry>


    @Query("UPDATE DashboardEntry SET dashboardEntryIndex = -1 " +
            " WHERE  dashboardEntryUid = :uid")
    abstract suspend fun pinEntry(uid: Long):Int

    @Query("UPDATE DashboardEntry SET dashboardEntryIndex = " +
            "(SELECT (SELECT MAX(dashboardEntryIndex) FROM DashboardEntry) +1) " +
            "WHERE dashboardEntryUid = :uid")
    abstract suspend fun unpinEntry(uid: Long):Int

    //TODO: Make it use a boolean argument
    @Query("UPDATE DashboardEntry SET dashboardEntryActive = 0 " +
            " WHERE dashboardEntryUid = :uid")
    abstract suspend fun deleteEntry(uid: Long):Int

    @Query("SELECT * FROM DashboardEntry WHERE dashboardEntryUid = :uid")
    abstract suspend fun findByUidAsync(uid: Long):DashboardEntry?

    @Update
    abstract suspend fun updateAsync(entity: DashboardEntry):Int

}
